File: src/brainfuck.zig
//! Brainfuck interpreter
const std = @import("std");

pub fn BrainFuck() type {
    return struct {
        const Self = @This();
        const maxTokens: u16 = 4096;
        const maxCells: u16 = 32;

        const BrainFuckError = error{
            TokensOverflow,
            TokensUnderflow,
            CellsOverflow,
            CellsUnderflow,
            JumpsError,
        };

        const Token = enum(u8) {
            stop = 0, // use as sentinel
            incPtr,
            decPtr,
            incCell,
            decCell,
            output,
            input,
            openBracket,
            closeBracket,
        };

        idx: u16 = 0,
        nestedJumps: u16 = 0,
        tokens: [maxTokens]Token = [_]Token{Token.stop} ** maxTokens,
        cells: [maxCells]u32 = [_]u32{0} ** maxCells,

        pub fn parse(self: *Self, code: []const u8) !void {
            var idx: u16 = 0;
            for (code) |c| {
                switch (c) {
                    '>' => self.tokens[idx] = Token.incPtr,
                    '<' => self.tokens[idx] = Token.decPtr,
                    '+' => self.tokens[idx] = Token.incCell,
                    '-' => self.tokens[idx] = Token.decCell,
                    '.' => self.tokens[idx] = Token.output,
                    ',' => self.tokens[idx] = Token.input,
                    '[' => {
                        self.tokens[idx] = Token.openBracket;
                    },
                    ']' => self.tokens[idx] = Token.closeBracket,
                    else => continue,
                }

                idx += 1;
                if (idx == maxTokens)
                    return BrainFuckError.TokensOverflow;
            }
        }

        pub fn execute(self: *Self) !void {
            var tokenIdx: u16 = 0;
            var cellIdx: u16 = 0;

            while (true) {
                switch (self.tokens[tokenIdx]) {
                    Token.stop => break,
                    Token.incPtr => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Increment data pointer by one (to point to the next cell to the right).
                        cellIdx += 1;
                        if (cellIdx == maxCells) return BrainFuckError.CellsOverflow;
                    },
                    Token.decPtr => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Decrement data pointer by one (to point to the next cell to the left).
                        if (cellIdx == 0) return BrainFuckError.CellsUnderflow;
                        cellIdx -= 1;
                    },
                    Token.incCell => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Increment the byte at the data pointer by one.
                        self.cells[cellIdx] += 1;
                    },
                    Token.decCell => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Decrement the byte at the data pointer by one.
                        self.cells[cellIdx] -= 1;
                    },
                    Token.output => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Output the byte at the data pointer.
                        if (self.cells[cellIdx] <= 255) {
                            const c: u8 = @truncate(self.cells[cellIdx]);
                            if (std.ascii.isPrint(c)) {
                                std.debug.print("{c}", .{c});
                            } else {
                                std.debug.print("<non printable u8 {}>", .{c});
                            }
                        } else {
                            std.debug.print("{}", .{self.cells[cellIdx]});
                        }
                    },
                    Token.input => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // Accept one byte of input, storing its value in the byte at the data pointer.
                        std.debug.print("Read byte not implemented\n", .{});
                    },
                    Token.openBracket => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // If the byte at the data pointer is zero, then instead of moving the instruction
                        // pointer forward to the next command, jump it forward to the command after the
                        // matching ] command.

                        // Check that nested jumps is equal to 0. Nested jump are managed here.
                        if (self.nestedJumps != 0)
                            return BrainFuckError.JumpsError;

                        if (self.cells[cellIdx] == 0) {
                            while (true) {
                                tokenIdx += 1;

                                if (tokenIdx == maxTokens)
                                    return BrainFuckError.JumpsError;

                                if (self.tokens[tokenIdx] == Token.closeBracket) {
                                    if (self.nestedJumps == 0) {
                                        break;
                                    } else {
                                        self.nestedJumps -= 1;
                                        continue;
                                    }
                                }

                                if (self.tokens[tokenIdx] == Token.openBracket)
                                    self.nestedJumps += 1;
                            }
                        } // else nothing to do
                    },
                    Token.closeBracket => {
                        //std.debug.print("Executing token {any} @ {}\n ", .{ self.tokens[tokenIdx], tokenIdx });
                        // If the byte at the data pointer is nonzero, then instead of moving the
                        // instruction pointer forward to the next command, jump it back to the command
                        // after the matching [ command.
                        if (self.nestedJumps != 0)
                            return BrainFuckError.JumpsError;

                        if (self.cells[cellIdx] != 0) {
                            while (true) {
                                tokenIdx -= 1;

                                if (tokenIdx == 0)
                                    return BrainFuckError.TokensUnderflow;

                                if (self.tokens[tokenIdx] == Token.openBracket) {
                                    if (self.nestedJumps == 0) {
                                        break;
                                    } else {
                                        self.nestedJumps -= 1;
                                        continue;
                                    }
                                }

                                if (self.tokens[tokenIdx] == Token.closeBracket)
                                    self.nestedJumps += 1;
                            }
                        } // else nothing to do
                    },
                }

                tokenIdx += 1;
                if (tokenIdx == maxTokens)
                    return BrainFuckError.TokensOverflow;
            }
        }
    };
}

File: src/brainfuck.rs
use std::collections::HashMap;

#[derive(Debug)]
enum Token {
    Incptr,
    Decptr,
    Incbyte,
    Decbyte,
    Outbyte,
    Inbyte,
    Forward,
    Backward,
}

pub struct Interpreter {
    ip: usize,                    // Instruction Pointer
    dp: usize,                    // Data Pointer
    cells: Vec<u8>,               // Vector of bytes
    insns: Vec<Token>,            // instrutions are list of Tokens
    jumps: HashMap<usize, usize>, // Keep track of jumps (forward and backwards)
}

// Our function doesn't return anything, it has just side effect
type Result<T> = std::result::Result<T, ()>;

impl Interpreter {
    pub fn new(code: &str) -> Result<Self> {
        let mut toks: Vec<Token> = Vec::new();

        code.chars().for_each(|c| match c {
            '>' => toks.push(Token::Incptr),
            '<' => toks.push(Token::Decptr),
            '+' => toks.push(Token::Incbyte),
            '-' => toks.push(Token::Decbyte),
            '.' => toks.push(Token::Outbyte),
            ',' => toks.push(Token::Inbyte),
            '[' => toks.push(Token::Forward),
            ']' => toks.push(Token::Backward),
            _ => {}
        });

        // Let's keep track of jumps in a second pass.
        let mut jumps_loc: Vec<usize> = Vec::new(); // keep track of open brackets position
        let mut jumps = HashMap::new();
        for (i, c) in toks.iter().enumerate() {
            match c {
                Token::Forward => {
                    jumps_loc.push(i);
                }
                Token::Backward => {
                    let forward_ip = match jumps_loc.pop() {
                        None => {
                            eprintln!("unbalanced brackets");
                            return Err(());
                        }
                        Some(ip) => ip,
                    };
                    jumps.insert(i, forward_ip);
                    jumps.insert(forward_ip, i);
                }
                _ => { // Nothing to do}
                }
            }
        }

        if !jumps_loc.is_empty() {
            eprintln!("Missing closed brackets");
            return Err(());
        }

        Ok(Self {
            ip: 0,
            dp: 0,
            insns: toks,
            cells: vec![0; 1024],
            jumps,
        })
    }

    fn interpreter_state(&self) {
        println!("-----------------------------------");
        println!("Next instruction: {:?}", self.insns[self.ip]);
        println!("IP: {:?}", self.ip);
        println!("DP: {:?}", self.dp);
        // print non empty cell
        for (id, c) in self.cells.iter().enumerate() {
            if *c != 0 {
                println!("cell[{:?}] = {:?}", id, *c);
            }
        }
    }

    pub fn run(&mut self, debug: bool) -> Result<()> {
        loop {
            if debug {
                self.interpreter_state()
            }

            // The program terminates when the instruction pointer
            // moves past the last command.
            if self.ip >= self.insns.len() {
                break;
            }

            match self.insns[self.ip] {
                Token::Incptr => {
                    self.dp += 1;
                    if self.dp >= self.cells.len() {
                        eprintln!("Memory overflow");
                        return Err(());
                    }
                }
                Token::Decptr => {
                    if self.dp == 0 {
                        eprintln!("Memory underflow");
                        return Err(());
                    }
                    self.dp -= 1;
                }
                Token::Incbyte => self.cells[self.dp] += 1,
                Token::Decbyte => self.cells[self.dp] -= 1,
                Token::Outbyte => {
                    print!("{:?}", self.cells[self.dp] as char);
                }
                Token::Inbyte => todo!("Inbyte"),
                Token::Forward => {
                    if self.cells[self.dp] == 0 {
                        match self.jumps.get(&self.ip) {
                            Some(new_ip) => self.ip = *new_ip, // IP is incremented at the end
                            None => {
                                eprintln!("Failed to match bracket");
                                return Err(());
                            }
                        }
                    }
                }
                Token::Backward => {
                    if self.cells[self.dp] != 0 {
                        match self.jumps.get(&self.ip) {
                            Some(new_ip) => self.ip = *new_ip, // IP is incremented at the end
                            None => {
                                eprintln!("Failed to match bracket");
                                return Err(());
                            }
                        }
                    }
                }
            }

            self.ip += 1;
        }

        println!();
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::brainfuck::Interpreter;

    #[test]
    pub fn github_profile() {
        let mut prog = Interpreter::new(
            "
            +++++ +++ [ >+++++ + [ >+>++>++<<<- ] >>+>++>+ [ < ] <- ] >>>-.>++++.<+.>++.--------.
            ",
        )
        .unwrap();
        prog.run(false).unwrap();
    }

    #[test]
    pub fn hello_test() {
        let mut prog = Interpreter::new(
            " source: wikipedia/Brainfuck
++++++++                Set Cell #0 to 8
[
    >++++               Add 4 to Cell #1; this will always set Cell #1 to 4
    [                   as the cell will be cleared by the loop
        >++             Add 2 to Cell #2
        >+++            Add 3 to Cell #3
        >+++            Add 3 to Cell #4
        >+              Add 1 to Cell #5
        <<<<-           Decrement the loop counter in Cell #1
    ]                   Loop until Cell #1 is zero; number of iterations is 4
    >+                  Add 1 to Cell #2
    >+                  Add 1 to Cell #3
    >-                  Subtract 1 from Cell #4
    >>+                 Add 1 to Cell #6
    [<]                 Move back to the first zero cell you find; this will
                        be Cell #1 which was cleared by the previous loop
    <-                  Decrement the loop Counter in Cell #0
]                       Loop until Cell #0 is zero; number of iterations is 8

The result of this is:
Cell no :   0   1   2   3   4   5   6
Contents:   0   0  72 104  88  32   8
Pointer :   ^

>>.                     Cell #2 has value 72 which is 'H'
>---.                   Subtract 3 from Cell #3 to get 101 which is 'e'
+++++++..+++.           Likewise for 'llo' from Cell #3
>>.                     Cell #5 is 32 for the space
<-.                     Subtract 1 from Cell #4 for 87 to give a 'W'
<.                      Cell #3 was set to 'o' from the end of 'Hello'
+++.------.--------.    Cell #3 for 'rl' and 'd'
>>+.                    Add 1 to Cell #5 gives us an exclamation point
>++.                    And finally a newline from Cell #6
            ",
        )
        .unwrap();
        prog.run(false).unwrap();
    }
}
